### PR TITLE
[RFC] Alternative IR approach

### PR DESCRIPTION
**Pull request by jacques-n**
An alternative serialized representation that I created a few months ago for a project I was working on, converted from Protobuf to Flatbuf for discussion here.

Relate to #10856 and #10934.

Key points of focus: 
- All field references are ordinals. Column names are not part of any of the core processing (beyond mapping from schema at input and output). This is really a concern of the presentation tool, not the execution/analysis layer. In Calcite, for example, the validator returns the field names of the actual return expression. The internal field names are meaningless (and actually cause user confusion).
- Leans heavily on concepts that Calcite has presented
- Projection is only used for calculation, not for column addition/removal. For the latter, each relational operation expresses whether it has a direct emit or a remapped emit.
- My thinking is this should support both logical and physical plans.
- The idea with function signatures is that public ones are organizationally namespaced and managed via a formal asset that the Arrow project contains. People can also make their own with a subsection of the namespace.
- As opposed to other representations, focus more on row-wise literals as 99% this representation will be easier to work with (e.g. a < 5).

### CODE REVIEW

File: format/IRFunction.fbs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// A unique identifier for a particular function definition. 
+table FunctionId {
+  // the function description identifier.
+  id: uint32;
+  
+  // The origin of the function definition. Should be mapped to a domain such as "org.apache.arrow"
+  // The github definition paths for definitions are defined in 
+  // github.com/apache/arrow/format/functions.txt. Org 0 == Apache Arrow canonical definitions.
+  // Organizations < 1B are canonical organizations. For private functions, use
+  // Org id > 1B
```
Comment (kkraus14) [inline format/IRFunction.fbs:31]: Is this 1 Byte or 1 Billion? How are canonical organizations defined outside of `org.apache.arrow`?

Comment (jacques-n) [reply format/IRFunction.fbs:31]: 1 billion. I was thinking that we'd have a doc on the Arrow git that manages the public orgs with pointers to those orgs definitions. Such as:
|Id  |Organization   |Path|
|---|---|---|
|0|org.apache.arrow|https://github.com/apache/arrow/base_sql_functions.yaml|
|1|org.bar.com|https://github.com/bar/repo1/the_missing_arrow_functions.yaml|
|2|org.foo.com|https://github.com/foo/repo1/stats_functions.yaml|
|3|org.random.com|https://random.com/playful_functions.yaml|

Comment (kkraus14) [reply format/IRFunction.fbs:31]: Is the idea to use something like this in the physical plan or logical plan or both?

Comment (jacques-n) [reply format/IRFunction.fbs:31]: I would expect both. I can come up with scenarios where you might want them namespaced (logical vs physical) but it feels like premature extensibility.

Comment (kkraus14) [reply format/IRFunction.fbs:31]: I'm concerned about having namespaced functions in the logical plan that aren't user defined functions.

For example, there's already functions and engines implemented in both C++ and Rust. If someone builds a logical plan from say a Python or R library, I imagine that logical plan should be able to be executed by either the C++ or Rust engine and that engine could choose which function implementation they should map to when building their physical plan.

I.E. say there's some arbitrary function that is standardized like `functionA` and there's both a C++ and Rust implementation of it. Would they have the same function IDs in different orgs or would they have the same org with different function IDs? How would an IR producer allow arbitrarily targeting either implementation?

**Comment by jacques-n** [comment]
Closing this in favor of collaborating on a cross-project approach at [substrait.io](https://substrait.io)

**Pull request closed without merging**